File: aoc2024-zig/src/challenges/day1.zig
const std = @import("std");
const util = @import("../util.zig");

const print = std.debug.print;

pub fn solve(input: []const u8, alloc: std.mem.Allocator) !void {
    var lines = util.lines(input);

    var left_column = std.ArrayList(u32).init(alloc);
    var right_column = std.ArrayList(u32).init(alloc);
    defer left_column.deinit();
    defer right_column.deinit();

    while (lines.next()) |line| {
        const left, const right = try util.splitOnceNumbers(u32, line, "   ");

        try left_column.append(left);
        try right_column.append(right);
    }

    std.mem.sort(u32, left_column.items, {}, std.sort.asc(u32));
    std.mem.sort(u32, right_column.items, {}, std.sort.asc(u32));

    var sum_part1: usize = 0;
    var sum_part2: usize = 0;
    for (left_column.items, right_column.items) |left, right| {
        sum_part1 += if (left > right) left - right else right - left;

        var count: usize = 0;
        for (right_column.items) |right_loop| {
            if (left == right_loop) count += 1;
        }
        sum_part2 += count * left;
    }

    print("Part 1: {}\n", .{sum_part1});
    print("Part 2: {}\n", .{sum_part2});
}

File: aoc2024-rust/src/challenges/day1.rs
pub fn solve(input: String) {
    let mut left_column: Vec<isize> = Vec::new();
    let mut right_column: Vec<isize> = Vec::new();
    for line in input.lines() {
        let mut iter = line.trim().split_whitespace();
        let left = iter.next().unwrap().parse::<isize>().unwrap();
        let right = iter.next().unwrap().parse::<isize>().unwrap();
        left_column.push(left);
        right_column.push(right);
    }
    left_column.sort();
    right_column.sort();

    let part1 = left_column
        .iter()
        .zip(right_column.iter())
        .fold(0, |acc, (left, right)| {
            acc + (if left > right {
                left - right
            } else {
                right - left
            })
        });

    let part2 = left_column.iter().fold(0, |acc, left| {
        acc + right_column.iter().filter(|right| left == *right).count() as isize * left
    });

    println!("Part 1: {}", part1);
    println!("Part 2: {}", part2);
}
